### ISSUE TITLE
glad_egl_find_core querying EGL_VERSION ahead of eglInitialize

### ISSUE THREAD

**Issue by tim-rex** [Dav1dde/glad#458]
It appears when `gladLoaderLoadEGL(NULL)` calls down to `glad_egl_find_core(NULL)`, it attempts to determine a suitable display, before then querying EGL_VERSION against it.

In the case where eglInitialize() has not previously been called (and in most scenarios, it likely would not have been), this causes Mesa to emit a debug warning (where `EGL_LOG_LEVEL=debug`) as follows:

```
libEGL debug: EGL user error 0x3001 (EGL_NOT_INITIALIZED) in eglQueryString
```

glad_egl_find_core shows:

```
static int glad_egl_find_core_{{ api|lower }}(EGLDisplay display) {
    int major, minor;
    const char *version;

    if (display == NULL) {
        display = EGL_NO_DISPLAY; /* this is usually NULL, better safe than sorry */
    }
    if (display == EGL_NO_DISPLAY) {
        display = eglGetCurrentDisplay();
    }
#ifdef EGL_VERSION_1_4
    if (display == EGL_NO_DISPLAY) {
        display = eglGetDisplay(EGL_DEFAULT_DISPLAY);
    }
#endif
#ifndef EGL_VERSION_1_5
    if (display == EGL_NO_DISPLAY) {
        return 0;
    }
#endif

    version = eglQueryString(display, EGL_VERSION);
    (void) eglGetError();

    if (version == NULL) {
        major = 1;
        minor = 0;
    } else {
        GLAD_IMPL_UTIL_SSCANF(version, "%d.%d", &major, &minor);
    }
```

Ultimately the egl_display being used here is EGL_DEFAULT_DISPLAY, and the EGL_VERSION is being returned by Mesa as 0, in which Glad then presume EGL 1.0

That seems appropriate, and I'd expect most (all?) implementations of EGL would similarly report null where eglInitialize()  has not already ocurred (perhaps not?).

I'm not sure that EGL_VERSION is intended to be queried ahead of eglInitialize() - which explains the rationale of the debug warning.

Would it be more appropriate for Glad to assume EGL version 1.0 where _display == NULL_ rather than attempting to retrieve the default display, and subsequently query GL_VERSION against it ?

I accept that there are very probably good reasons for the current behaviour, and this issue is really just splitting hairs (it is inconsequential, and easily ignored)

Also, thank you for creating and maintaining this fantastic (and very useful) library. 

Reference: https://gitlab.freedesktop.org/mesa/mesa/-/issues/10710

**Comment by Dav1dde** [comment]
I am wondering if instead of querying the version via `eglQueryString` it would be okay to just call `eglInitialize` which also returns the version.

> I'm not sure that EGL_VERSION is intended to be queried ahead of eglInitialize() - which explains the rationale of the debug warning.

No, it even generates an error and returns `EGL_NOT_INITIALIZED`, that's why there is the call to `(void) eglGetError();` right after to clear the error. Which is also the warning you're seeing.

> That seems appropriate, and I'd expect most (all?) implementations of EGL would similarly report null where eglInitialize() has not already ocurred (perhaps not?).

The spec at least requires it, but wouldn't be the first time that there is an implementation which doesn't follow the spec.

> Would it be more appropriate for Glad to assume EGL version 1.0 where display == NULL rather than attempting to retrieve the default display, and subsequently query GL_VERSION against it ?

Maybe, but I don't want to introduce a breaking change here which is especially really subtle. For convenience glad always allowed you to pass `NULL` and it would just fall back to the current or default display. Possible that is is actually something people aren't using or shouldn't be, but I think it's too late to change.

**Comment by tim-rex** [comment]
That sounds more than fair, it doesn't seem necessary to introduce a possibly breaking change, thank you for responding so comprehensively.

**Comment by planetmarshall** [comment]
> 
> Maybe, but I don't want to introduce a breaking change here which is especially really subtle. For convenience glad always allowed you to pass `NULL` and it would just fall back to the current or default display. Possible that is is actually something people aren't using or shouldn't be, but I think it's too late to change.
> 

Sorry to revisit an old thread - but if Glad always resolves the version to 1.0 (this is the behaviour I am seeing, You don't get a correct version until egllnitialize has been called) - then the entrypoints for subsequent versions will not be loaded.


Currently in order to bootstrap the EGL loading process I have to do something like:

```.cpp
auto egl_version = gladLoaderLoadEGL(NULL); // always returns 10000 because egl is not initialized when the version is queried
    
auto egl_display = eglGetDisplay(EGL_DEFAULT_DISPLAY);
auto egl_major = EGLint();
auto egl_minor = EGLint();
eglInitialize(egl_display, &egl_major, &egl_minor);
egl_version = gladLoaderLoadEGL(egl_display); // now loads the correct version and the additional entrypoints are available
```
